### PR TITLE
Add ACCESS-OM3 builder

### LINKED ISSUE #151: ACCESS-OM3 Builder
# Is your feature request related to a problem? Please describe.

To support quick analysis of new runs in access-om3, we will need a access-om3 builder. 

# Describe the feature you'd like

We want to be able to build a temporary / adhoc datastore from a single run of access-om3 output for quick analysis against big picture metrics (e.g. similar to the figures in access-om2 paper)

# Describe alternatives you've considered

We could use file paths until access-om3 is further developed

# Additional context

At some point we might automate building the datastore into payu.

### PR DESCRIPTION
**Pull request by dougiesquire**
This PR adds a builder class for ACCESS-OM3 data. Closes #151

### CODE REVIEW

**Comment by dougiesquire** [comment]
Thanks for agreeing to review @anton-seaice. Don't worry about all the new files in `tests/data` - these are just stripped-back real OM3 output that are used in some of the tests. The tests themselves should make it clear where/how these are used.

**Review by anton-seaice — COMMENTED** [review]
This seems to work like expected (https://github.com/anton-seaice/sandbox/blob/171f8103d0140d8b98e25637e500d2b8501ef00c/cat_om3_builder.ipynb)

I made a datastore using a CICE6+MOM6 config, and another usign CICE6+MOM6+WW3 and they appear to load, have variables and can access the data etc

File: docs/datastores/builders.rst [hunk 1]
```diff
@@ -39,8 +39,15 @@ ACCESS-CM2 output: :code:`AccessCm2Builder`
    :special-members: __init__, build, save
    :noindex:
 
+ACCESS-OM3 output: :code:`AccessOm3Builder`
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. autoclass:: access_nri_intake.source.builders.AccessOm3Builder
+   :special-members: __init__, build, save
+   :noindex:
+
```
[the remark this answers is not in the document]
Comment (dougiesquire) [reply docs/datastores/builders.rst:48]: This rendered successfully locally for me, so let's suck it and see.

File: src/access_nri_intake/source/builders.py [hunk 2]
```diff
@@ -266,6 +266,91 @@ def parser(file):
...
+                start_date,
+                end_date,
+                variable_list,
+                variable_long_name_list,
+                variable_standard_name_list,
+                variable_cell_methods_list,
+                variable_units_list,
+            ) = parse_access_ncfile(file)
+
+            if ("mom6" in filename) or ("ww3" in filename):
```
[the remark this answers is not in the document]
Comment (dougiesquire) [reply src/access_nri_intake/source/builders.py:326]: Good question. I don't know... The [allowable vocab](https://github.com/ACCESS-NRI/schema/blob/355bc87ac44fdc88ac4f19e7344fe48e67fb9e4e/experiment_asset.json#L27-L48) we have for realms atm follows CMIP6, which doesn't include a specific "waves" realm. We could add one, but that would require updating the schema, which is a little more work. The easiest route is just to say that ocean waves are part of the ocean (which is true I guess...).

But if you feel strongly, I'd be happy to add a `wave` realm?

[the remark this answers is not in the document]
Comment (dougiesquire) [reply src/access_nri_intake/source/builders.py:326]: The obvious advantage to having a `wave` realm is that you'd be able to filter the catalog for only experiments waves

**Comment by anton-seaice** [comment]
Screenshot 2024-03-01 at 1 21 24 pm

In my WW data in the datastore, there is no "start_date" & "end_date" ?
Screenshot 2024-03-01 at 1 22 29 pm

File: src/access_nri_intake/source/builders.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (anton-seaice) [reply src/access_nri_intake/source/builders.py:326]: I don't feel strongly at all, it feels like bigger question than this review!

**Comment by dougiesquire** [comment]
> In my WW data in the datastore, there is no "start_date" & "end_date" ?

Yeah, the WW3 data contains no time bounds info, so when there is only one time step per file there's no way to determine what the frequency is and it assumes fixed frequency. I could still add strings for the start and end dates in this case? In which case we'd have something like:

| frequency | start_date | end_date |
| ---------- | ---------- | ---------- |
| fx | 1958-01-02, 00:00:00 | 1958-01-02, 00:00:00 |

**Comment by anton-seaice** [comment]
> Yeah, the WW3 data contains no time bounds info, so when there is only one time step per file there's no way to determine what the frequency is and it assumes fixed frequency. I could still add strings for the start and end dates in this case?...

Does it make more sense to add this to the WW3 output?

**Comment by dougiesquire** [comment]
> Does it make more sense to add this to the WW3 output?

It would _definitely_ be better to have this in the output, but that requires some automated postprocessing (unless it can be configured in WW3?). These aren't the first files to be missing time bounds and they won't be the last so some sensible default behaviour is probably still needed

**Comment by anton-seaice** [comment]
> > Does it make more sense to add this to the WW3 output?
> 
> It would _definitely_ be better to have this in the output, but that requires some automated postprocessing (unless it can be configured in WW3?). These aren't the first files to be missing time bounds and they won't be the last so some sensible default behaviour is probably still needed

It doesn't look like its configurable. A code change shouldn't be too hard though. We will need to dome something for time frequency eventually because at the moment there is no way to know if the output is daily or monthly (or something else), except for looking in `nuopc.runconfig`

Adding some sort of start and end date handling should improve searching.

**Comment by dougiesquire** [comment]
> Adding some sort of start and end date handling should improve searching.

With https://github.com/ACCESS-NRI/access-nri-intake-catalog/pull/152/commits/1f613a79ec6baced0bfcb34088d6ef58f52e81a2, we now parse the start and end dates from the WW3 files (but they are the same...)

**Comment by anton-seaice** [comment]
This seems to work now. We've added a different annoyance, which is the mom static file now produces a 'start_date' and 'end_date' in the catalog. But that is due to a quirk of the source file, not this builder :)

**Comment by anton-seaice** [comment]
I raised https://github.com/COSIMA/access-om3/issues/110 to consider the missing time_bounds in WW output

File: src/access_nri_intake/source/builders.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (anton-seaice) [reply src/access_nri_intake/source/builders.py:326]: deferring to https://github.com/ACCESS-NRI/access-nri-intake-catalog/issues/155

**Comment by dougiesquire** [comment]
> This seems to work now. We've added a different annoyance, which is the mom static file now produces a 'start_date' and 'end_date' in the catalog. But that is due to a quirk of the source file, not this builder :)

Yeah, there isn't really anything we can do about data that's missing metadata or saved in a strange way. I think adding the start and end date whenever there is a time axis is the most defensible approach, otherwise we might be throwing away potentially useful info

**Review by anton-seaice — APPROVED** [review]

**Comment by dougiesquire** [comment]
Thanks for a thorough review @anton-seaice

**Pull request merged**